cpp/src/arrow/python/builtin_convert.cc

```diff
@@ -722,25 +736,60 @@ class ListConverter : public TypedConverterVisitor<ListBuilder, ListConverter> {
  public:
   Status Init(ArrayBuilder* builder) override;
 
-  Status AppendItem(const OwnedRef& item) {
+  Status AppendItem(PyObject* obj) override {
     RETURN_NOT_OK(typed_builder_->Append());
-    PyObject* item_obj = item.obj();
-    const auto list_size = static_cast<int64_t>(PySequence_Size(item_obj));
-    return value_converter_->AppendData(item_obj, list_size);
+    const auto list_size = static_cast<int64_t>(PySequence_Size(obj));
+    return value_converter_->AppendMultiple(obj, list_size);
   }
 
  protected:
   std::shared_ptr<SeqConverter> value_converter_;
 };
 
+class StructConverter : public TypedConverterVisitor<StructBuilder, StructConverter> {
+ public:
+  Status Init(ArrayBuilder* builder) override;
+
+  Status AppendItem(PyObject* obj) override {
+    RETURN_NOT_OK(typed_builder_->Append());
+    if (!PyDict_Check(obj)) {
+      return Status::TypeError("dict value expected for struct type");
+    }
+    // NOTE we're ignoring any extraneous dict items
+    for (int i = 0; i < num_fields_; i++) {
+      PyObject* nameobj = PyList_GET_ITEM(field_name_list_.obj(), i);
+      PyObject* valueobj = PyDict_GetItem(obj, nameobj);  // borrowed
+      RETURN_IF_PYERROR();
+      RETURN_NOT_OK(value_converters_[i]->AppendSingle(valueobj ? valueobj : Py_None));
```

Comment:
I'm not sure how much of a measurable performance impact it will have (it would be good to know for future reference), but I made a separate vector of raw pointers to the internal builders for `RecordBatchBuilder` to avoid the extra shared_ptr overhead on the inner loop https://github.com/apache/arrow/blob/master/cpp/src/arrow/table_builder.h#L105

Reply:
I'll give a try.

Reply:
Ok, I tried it with the following micro-benchmark:
```
$ python -m timeit -s "import pyarrow as pa; ty=pa.struct([pa.field('x', pa.int32())]); data=[None]*1000000" "pa.array(data, type=ty)"
```

* unpatched: 49.3 msec per loop
* with unique_ptr instead of shared_ptr: 44.3 msec per loop
* with raw pointers in addition to shared_ptr: 43.6 msec per loop

Since an additional array of raw pointers adds some complication, I'm tempted to go with the unique_ptr solution (which AFAICS looks sane).

Reply:
SGTM

Reply:
Ok, I migrated the code to unique_ptr.